react-components/src/components/NodeCacheProvider/FdmNodeCache.ts

```diff
@@ -28,36 +31,139 @@ import { partition } from 'lodash';
 
 import assert from 'assert';
 import { fetchNodesForNodeIds } from './requests';
-
-export type ModelRevisionKey = `${number}-${number}`;
-export type ModelRevisionToEdgeMap = Map<ModelRevisionKey, FdmEdgeWithNode[]>;
+import { type ThreeDModelMappings } from '../../hooks/types';
 
 export class FdmNodeCache {
-  private readonly _revisionNodeCaches = new Map<RevisionKey, RevisionFdmNodeCache>();
+  private readonly _revisionNodeCaches = new Map<ModelRevisionKey, RevisionFdmNodeCache>();
 
   private readonly _cdfClient: CogniteClient;
   private readonly _fdmClient: FdmSDK;
 
-  private readonly _completeRevisions = new Set<RevisionKey>();
+  private readonly _completeRevisions = new Set<ModelRevisionKey>();
 
   public constructor(cdfClient: CogniteClient, fdmClient: FdmSDK) {
     this._cdfClient = cdfClient;
     this._fdmClient = fdmClient;
   }
 
+  public async getMappingsForFdmIds(
+    externalIds: DmsUniqueIdentifier[],
+    modelRevisionIds: ModelRevisionId[]
+  ): Promise<ThreeDModelMappings[]> {
+    const [cachedModelRevisionIds, nonCachedModelRevisionIds] = partition(
+      modelRevisionIds,
+      (id) => {
+        const revisionKey = createModelRevisionKey(id.modelId, id.revisionId);
+        return this._completeRevisions.has(revisionKey);
+      }
+    );
+
+    const nonCachedModelMappings = await this.getNonCachedModelMappings(
+      externalIds,
+      nonCachedModelRevisionIds
+    );
+
+    const cachedModelMappings = this.getCachedModelMappings(cachedModelRevisionIds, externalIds);
+
+    const combinedList = [...cachedModelMappings, ...nonCachedModelMappings];
+    return combinedList;
+  }
+
+  private getCachedModelMappings(
+    modelRevisionIds: ModelRevisionId[],
+    externalIds: DmsUniqueIdentifier[]
+  ): ThreeDModelMappings[] {
+    const inputExternalIdSet = new Set<FdmKey>(
+      externalIds.map((id) => createFdmKey(id.space, id.externalId))
+    );
+
+    return modelRevisionIds.map((modelRevisionId) => {
+      return this.getCachedModelMappingForRevision(modelRevisionId, inputExternalIdSet);
+    });
+  }
+
+  private getCachedModelMappingForRevision(
+    modelRevisionId: ModelRevisionId,
+    relevantFdmKeySet: Set<FdmKey>
+  ): ThreeDModelMappings {
+    const revisionCache = this.getOrCreateRevisionCache(
+      modelRevisionId.modelId,
+      modelRevisionId.revisionId
+    );
+
+    const relevantCachedEdgeData = intersectWithStartNodeIdSet(
+      revisionCache.getAllEdges(),
+      relevantFdmKeySet
+    );
+
+    const mappings = new Map(
+      relevantCachedEdgeData.map((data) => [data.edge.startNode.externalId, data.node])
+    );
+
+    return {
+      ...modelRevisionId,
+      mappings
+    };
+  }
+
+  private async getNonCachedModelMappings(
+    uniqueIds: DmsUniqueIdentifier[],
+    modelRevisions: ModelRevisionId[]
+  ): Promise<ThreeDModelMappings[]> {
+    if (modelRevisions.length === 0) {
```

Comment:
Don't we need `uniqueIds.length` check along with `modelRevision.length`?

Reply:
We don't really _need_ it.. The reason I check for `modelRevision`, is that this function is only run for non-cached modelRevisions, which is none, in many cases. `uniqueIds` will only be zero if the user didn't supply any, which I think is uncommon.

But whatever, it's more efficient to put it there anyway.